### DISCUSSION TITLE
Implement local geocoding or allow choosing different geocoding provider

### DISCUSSION THREAD

**Discussion by ghost** [breezy-weather/breezy-weather#1305] [Ideas]
### Describe your idea

Geocoding is not always supported by every source, notably the more libre sources such as Open-Meteo and OpenWeather, this gives users who prefer more open weather sources a "disadvantage" or inconvenience factor.

I suggest either implementing some kind of geocoding method locally on device to display where the Current location was last updated to, or allow choosing a different source that supports geocoding just for the geocoding function to work.

I personally would only like to have the geocoding feature to see if the Current location is correct or not, whether I need to update it by turning on Location data or not. This would be useful for users who don't have Location permission on by default all the time.

### Acknowledgements

- [X] I have searched the [existing issues](https://github.com/breezy-weather/breezy-weather/issues) and [existing discussions](https://github.com/breezy-weather/breezy-weather/discussions/categories/ideas) and this is a new ticket, **NOT** a duplicate or related to another open OR closed issue or discussion.
- [X] The idea I’m suggesting is not in the [list of features that will not be implemented](https://github.com/breezy-weather/breezy-weather/blob/main/README.md#features-that-will-not-be-implemented).
- [X] I have updated the app to **[version 5.2.8 or later](https://github.com/breezy-weather/breezy-weather/releases/latest)**. Even if I’m getting updates from F-Droid default repo, I don’t need to uninstall/reinstall to update.
- [X] I wrote my request in English, and any screenshot I attached are also in English (language can be changed from Settings > Appearance). I can write in my native language under the English text, if necessary.
- [X] I understand that this app is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY, so I may not get a positive reply to my request, if any at all.

**Comment by papjul** [comment]
> implementing some kind of geocoding method locally on device

We already have an offline reverse geocoder able to detect the country from longitude/latitude. Do you suggest you prefer to see "United States" rather than "Current location"? I think "Current location" is more detailed than "United States".

> allow choosing a different source that supports geocoding just for the geocoding function to work

That would be a duplicate of #1258, right?

> see if the Current location is correct or not

That would be a 3rd different request. Reverse geocoding does not tell you if the location is correct, it's a mapping to an address than can itself be incorrect, while the actual location can be correct. So what you would like is rather being able to see on a map where you were last located, right? That's something I've been thinking of, and that could be implemented after the map library is added for feature #10 

> whether I need to update it by turning on Location data or not

If you want a location based on a specific point on a map that does not rely on location permission, you might be interested in #10 

____

Please decide between the different requests which one you want to focus this discussion on, and if multiple requests, please open other discussions for each ;) Thanks!

**Reply by ghost** [reply]
> We already have an offline reverse geocoder able to detect the country from longitude/latitude. Do you suggest you prefer to see "United States" rather than "Current location"? I think "Current location" is more detailed than "United States".

"United Statues" or "Current geolocation" does not tell whether the city or sub-region of a country the weather data is based is correct or not. I would prefer US: Florida or some kind of approximate city. My main issue is not knowing what city the data is based on so I don't know when to change it. It would be great to know the exact city. When the app displays "Current location" does it pull the weather data to the collective "United States"? If so it would still be better to see "United States", but I would prefer a more accurate geocode if possible.

> That would be a duplicate of #1258, right?

It seems like it but accessible to all regions not just chine.

> That would be a 3rd different request. Reverse geocoding does not tell you if the location is correct, it's a mapping to an address than can itself be incorrect, while the actual location can be correct.

That's fine, the user (hopefully) knows where they are so even if the exact location/address is wrong, less accurate, it's not that big of a deal since weather data is mostly the same in an entire city, I would believe. I just think users would prefer to see the name of city so if they were in a different region 2 hours ago they wouldn't see the wrong weather data without knowing it is from a different city.

> So what you would like is rather being able to see on a map where you were last located, right? That's something I've been thinking of, and that could be implemented after the map library is added for feature #10

I don't think something like that is necessary just purely from a maintanance, code burden perspective. A simple string like "Current location: Near <Insert City Name>" would be fine but of course a map can also show whether the Current location data needs refreshing or not even more accurately too.

> If you want a location based on a specific point on a map that does not rely on location permission, you might be interested in #10

Wouldn't it just do two things? 1: display all the location sources on a map (Current location and city locations) 2: adds the ability to manually add a location. Really cool and would solve my idea but first wouldn't it be easier to implement the text-only solution?

So just to sum it up: per my usecase I just want to see what location/city the "Current location" is basing it's weather data on.

**Reply by papjul** [reply]
> > We already have an offline reverse geocoder able to detect the country from longitude/latitude. Do you suggest you prefer to see "United States" rather than "Current location"? I think "Current location" is more detailed than "United States".
> 
> "United Statues" or "Current geolocation" does not tell whether the city or sub-region of a country the weather data is based is correct or not. I would prefer US: Florida or some kind of approximate city.

Not possible with the offline reverse geocoder, adding subdivisions of each country would make the app way too big for just a cosmetic feature.

> My main issue is not knowing what city the data is based on so I don't know when to change it. It would be great to know the exact city. When the app displays "Current location" does it pull the weather data to the collective "United States"? If so it would still be better to see "United States", but I would prefer a more accurate geocode if possible.

Aaaah, we have a misunderstanding here. The weather data is based on longitude/latitude only, not on city, country or something else. The reverse geocoding serves no other purpose than being a more readable form of the longitude and latitude (so it's just cosmetic).

> > That would be a duplicate of #1258, right?
> 
> It seems like it but accessible to all regions not just chine.

I don't plan on making it a China-specific feature.

> > That would be a 3rd different request. Reverse geocoding does not tell you if the location is correct, it's a mapping to an address than can itself be incorrect, while the actual location can be correct.
> 
> That's fine, the user (hopefully) knows where they are so even if the exact location/address is wrong, less accurate, it's not that big of a deal since weather data is mostly the same in an entire city, I would believe. I just think users would prefer to see the name of city so if they were in a different region 2 hours ago they wouldn't see the wrong weather data without knowing it is from a different city.

What I suggest is not for the user to know where they are, but to know where they were last located by Android at Breezy Weather request. Knowing you're in New York City will not be helpful if you crossed the whole city during the day; you won't know if "New York City" means far west of the city, or far east of the city, while a point on the map will tell you straight away ;) 

> > So what you would like is rather being able to see on a map where you were last located, right? That's something I've been thinking of, and that could be implemented after the map library is added for feature #10
> 
> I don't think something like that is necessary just purely from a maintanance, code burden perspective. A simple string like "Current location: Near " would be fine but of course a map can also show whether the Current location data needs refreshing or not even more accurately too.

Alright, let's focus on the other options then ;)

> > If you want a location based on a specific point on a map that does not rely on location permission, you might be interested in #10
> 
> Wouldn't it just do two things? 1: display all the location sources on a map (Current location and city locations) 2: adds the ability to manually add a location. Really cool and would solve my idea but first wouldn't it be easier to implement the text-only solution?

There are no plans to do "1" in #10, only "2".
We can do both; I can't tell you which one of the two will be implemented first. Yes, reverse geocoding is probably easier to do, but the map probably covers more cases and is probably the most interesting from a priority perspective.

> So just to sum it up: per my usecase I just want to see what location/city the "Current location" is basing it's weather data on.

The weather data is based solely on longitude and latitude (hence, my suggestion for a map) since forecast is generated on a grid, except for AccuWeather which uses a mix of city + lon/lat (but in that case, you already have a working reverse geocoder).

So, if I sum up, you either want:
- to be able to choose a reverse geocoding source different from the main source of a location (especially when main source has no reverse geocoding feature) => that would be #1258 
- to add a location that is a fixed point on the map, that would not be tied to the Android GPS => that would be #10 

If I understood correctly, can we close this as duplicate of #10 or #1258?

**Reply by ghost** [reply]
> Not possible with the offline reverse geocoder, adding subdivisions of each country would make the app way too big for just a cosmetic feature.

Wouldn't it be possible to use some kind of map data and approximate which city the user is closer to?

> What I suggest is not for the user to know where they are, but to know where they were last located by Android at Breezy Weather request.

That's the issue the user may not remember that.

> * to be able to choose a reverse geocoding source different from the main source of a location (especially when main source has no reverse geocoding feature) => that would be [New secondary source type: reverse geocoding #1258](https://github.com/breezy-weather/breezy-weather/discussions/1258)

That is one solution yes.

> * to add a location that is a fixed point on the map, that would not be tied to the Android GPS => that would be [Add location from a map #10](https://github.com/breezy-weather/breezy-weather/issues/10)

That would also be cool but not my initial suggestion.

I would prefer if the app could approximate by itself what city, parts of the city the user is closest to/inside of. Could the app somehow invoke OSM data with only the region data or something like that, so US>New York>Bronx can be approximated. Would that data still make the app too big?  The app would only need to fetch a specific countries map data in this case.

> There are no plans to do "1" in #10, only "2".
> We can do both; I can't tell you which one of the two will be implemented first. Yes, reverse geocoding is probably easier to do, but the map probably covers more cases and is probably the most interesting from a priority perspective.

I just assumed because in the #10 issue there were multiple cities marked on the image and if you'd want to solve this issue with the map, you'd need to display the Current location coordinates on the map too to indicate whether that location is up to date or not.

I'm getting kind of caught up with the details so I would just lay down how I see it:

If approximating by text is not possible then the map option would be an ok solution if: there will be some kind of pin or location indicator for Current location coordinates, there will be a string that says when was the last time Current location GPS data was updated.

It would also be cool if the the manual city locations and manual coordinate locations were visible on this map with the ability to add, delete and modify these locations.

**Reply by papjul** [reply]
> > Not possible with the offline reverse geocoder, adding subdivisions of each country would make the app way too big for just a cosmetic feature.
> 
> Wouldn't it be possible to use some kind of map data and approximate which city the user is closer to?

Yes, that's the definition of "reverse geocoding", but doing that offline is not realistic, all polygons, all translations, etc, of all cities in the world. The GeoNames database weights 383 MB zipped. Our app weights 10 MB in comparison, including a 2 MB unzipped country database. Can you imagine your phone going through all cities in the world to find which one is the closest to you? That would take hours. An online service is optimized to process that kind of request.

> > What I suggest is not for the user to know where they are, but to know where they were last located by Android at Breezy Weather request.
> 
> That's the issue the user may not remember that.

What do you mean?

> I would prefer if the app could approximate by itself what city, parts of the city the user is closest to/inside of. Could the app somehow invoke OSM data with only the region data or something like that, so US>New York>Bronx can be approximated. Would that data still make the app too big? The app would only need to fetch a specific countries map data in this case.

TimeZone is a mandatory field for a reverse geocoder; we can't use OSM API. And we are not going to set up an overly complex offline system that can download country-specific data, unzip it, update it when there is an update, etc. I'm also not sure OSM would be happy we use their bandwidth when there is an API exactly for that purpose. And I'm not even sure it would not still take hours (or at least minutes) to loop through all cities of the United States to find the closest one.

> > There are no plans to do "1" in #10, only "2".
> > We can do both; I can't tell you which one of the two will be implemented first. Yes, reverse geocoding is probably easier to do, but the map probably covers more cases and is probably the most interesting from a priority perspective.
> 
> I just assumed because in the #10 issue there were multiple cities marked on the image and if you'd want to solve this issue with the map, you'd need to display the Current location coordinates on the map too to indicate whether that location is up to date or not.

The screenshot you see is to illustrate the "search to add a new location" feature, not to show currently added locations (I have no plans to do that). I only suggested showing somewhere else in the app the current location specifically on a map, but you said it was not necessary.

> I'm getting kind of caught up with the details so I would just lay down how I see it:
> 
> If approximating by text is not possible then the map option would be an ok solution if: there will be some kind of pin or location indicator for Current location coordinates, there will be a string that says when was the last time Current location GPS data was updated.

That's not what I suggested. I suggested adding a fixed location. You choose your longitude and latitude by pinpointing on the map, so you have a location specifically in your neighborhood instead of the center of the city if you were to use the search, which I believed was the reason you used "current location" then turned off location permission? However, that location would not change with the GPS, it would remain where you decided to put it.

> It would also be cool if the the manual city locations and manual coordinate locations were visible on this map with the ability to add, delete and modify these locations.

So, in the end, you want to see all your existing locations on a map? I wasn't thinking about doing something like that.
If "modify" means "changing the coordinates of a manually added location" the answer is no for technical reasons. If you want different coordinates, you need to delete and add a new different location.

____

Offline geocoding of cities being excluded, what would be the one solution for you? Isn't #1258 enough?
